src/qmltypes/qmlfilter.cpp

```diff
@@ -770,27 +763,121 @@ int QmlFilter::keyframeIndex(Mlt::Animation &animation, int position)
     return result;
 }
 
+void QmlFilter::startUndoTracking()
+{
+    m_previousState = Mlt::Properties();
+    m_previousState.inherit(m_service);
+    if (!m_previousState.property_exists(kShotcutAnimInProperty)) {
+        m_previousState.set(kShotcutAnimInProperty, 0);
+    }
+    if (!m_previousState.property_exists(kShotcutAnimOutProperty)) {
+        m_previousState.set(kShotcutAnimOutProperty, 0);
+    }
+}
+
+void QmlFilter::startChangeParameterCommand(const QString &desc)
+{
+    if (!m_previousState.count()) {
+//        LOG_DEBUG() << "Undo tracking has not started yet";
+        return;
+    }
+    m_changeInProgress++;
```

Comment:
I found that the filter changes can end up nesting start commands. For example, modifyig a parameter might trigger a callback that will modify other parameters. So the changeInProgress flag is really a counter to keep track of when the nested calls have been unrolled.